nemo/collections/multimodal/speech_llm/data/build_dataset.py

```diff
@@ -0,0 +1,233 @@
+                else model_instance.perception.asr_model.tokenizer
+            )
+            # get the canary dataset as a processor to read canary prompt from data and convert_canary_prompt_to_text
+            canary_processer = PromptedAudioToTextLhotseDataset(
+                tokenizer=perception_tokenizer,
+                prompt_format_fn=get_prompt_format_fn('canary'),
+                inference=True,
+            )
+        else:
+            canary_processer = None
+        return LhotseAudioQuestionAnswerDataset(
+            tp,
+            default_question="answer the question according to the previous audio",
+            tokens_to_generate=data_cfg.get('tokens_to_generate', 0),
+            pad_to_max_length=data_cfg.get('pad_to_max_length', False),
+            max_seq_length=data_cfg["max_seq_length"],
+            canary_processor=canary_processer,
+            convert_canary_prompt_to_text=data_cfg.get('convert_canary_prompt_to_text', False),
+            prepend_to_exist_question=data_cfg.get('prepend_to_exist_question', None),
+            canary_tokens_augment_ratio=data_cfg.get('canary_tokens_augment_ratio', 0.0),
+            random_context_prob=data_cfg.get('random_context_prob', 0.0),
+            random_context_positive_percent=data_cfg.get('random_context_positive_percent', 0.1),
+        )
+
+    # Notably, the data weights are controlled by either bucketing_weights
+    # or concat_sampling_probabilities depending on the dataset type.
+    if data_cfg.get('is_tarred', False):
+        return get_tarred_audio_text_dataset_from_config(
+            config=data_cfg,
+            tokenizer=model_instance.tokenizer,
+            augmentor=augmentor,
+            sep_id=model_instance.sep_id,
+            answer_only_loss=model_instance.cfg.get('answer_only_loss', True),
+            virtual_tokens=model_instance.virtual_tokens,
+            global_rank=parallel_state.get_data_parallel_rank(),
+            world_size=parallel_state.get_data_parallel_world_size(),
+        )
+    else:
+        return get_audio_text_dataset_from_config(
+            manifest_filepath=data_cfg.manifest_filepath,
+            config=data_cfg,
+            tokenizer=model_instance.tokenizer,
+            augmentor=augmentor,
+            is_train=is_train,
+            sep_id=model_instance.sep_id,
+            answer_only_loss=model_instance.cfg.get('answer_only_loss', True),
+            virtual_tokens=model_instance.virtual_tokens,
+        )
+
+
+def build_salm_dataloader(dataset, data_cfg, consumed_samples=0, is_predict=False, is_eval=False):
+    """Buld dataloader given an input dataset."""
+    if data_cfg.get("use_lhotse"):
+        from nemo.collections.common.data.lhotse import get_lhotse_dataloader_from_config
+
+        # for eval, we need to create separate dataset so as to report splitted numbers
+        if data_cfg.get('is_tarred', False) or (is_eval == False and is_predict == False):
```

Comment:
you shouldn't rely on the flag `is_tarred` for choosing lhotse dataloader, lhotse supports more formats than nemo json and nemo tar and auto-deduces `is_tarred`

Reply:
good catch! Done.